sql: implement pgvector datatype and evaluation

jordanlewis opened the pull request:
This commit adds the pgvector datatype and associated evaluation operators and functions. It doesn't include index acceleration.

Functionality included:

- `CREATE EXTENSION vector`
- `vector` datatype with optional length, storage and retrieval in non-indexed table columns
- Equality and inequality operators
- `<->` operator - L2 distance
- `<#>` operator - (negative) inner product
- `<=>` operator - cosine distance
- `l1_distance` builtin
- `l2_distance` builtin
- `cosine_distance` builtin
- `inner_product` builtin
- `vector_dims` builtin
- `vector_norm` builtin

Updates #121432
Epic: None

Release note (sql change): implement pgvector encoding, decoding, and operators, without index acceleration.

jordanlewis committed:
sql: implement pgvector datatype and evaluation

Release note (sql change): implement pgvector encoding, decoding, and
operators, without index acceleration.

DrewKimball reviewed (commented):
This is so cool!

We should also do some mixed-version testing to make sure the new type is correctly disallowed until a cluster is upgraded ([example](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/mixed_version_sequence_per_node_cache)).

> Reviewed 54 of 68 files at r1, 21 of 21 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-:-Ny82EQT46uZNjYVzjrc:brbkktw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @nameisbhaskar, @nkodali, @rharding6373, @vidit-bhat, and @yuzefovich)

___
*[`pkg/sql/opt/ops/scalar.opt` line 538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7hBdH3chbzT3cCown:-Ny7hBdH3chbzT3cCowo:brmqeux) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/opt/ops/scalar.opt#L538)):*
> ```
>     Left ScalarExpr
>     Right ScalarExpr
> }
> ```

Any particular reason why we aren't also doing the l1 distance `<+>`? pg_vector seems to support it: https://github.com/pgvector/pgvector/blob/258eaf58fdaff1843617ff59ea855e0768243fe9/README.md?plain=1#L142-L147
___
*[`pkg/sql/parser/sql.y` line 14277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7baQNA3H3KC7vw6Qf:-Ny7baQO38ZZ8Verz5mM:b-e9kxpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/parser/sql.y#L14277)):*
> ```Yacc
>   }
> 
> const_vector:
> ```

Can we add some parser tests?
___
*[`pkg/sql/scanner/scan.go` line 338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7e11DDpYZKMoTxH87:-Ny7e11DDpYZKMoTxH88:b-m4bfjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/scanner/scan.go#L338)):*
> ```Go
> 				return
> 			}
> 			return
> ```

Here and below, we're advancing the position of the scanner without setting the token ID. Should we backtrack if it isn't a vector operator?
___
*[`pkg/sql/sem/builtins/pgvector_builtins.go` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7gH-e36hk2xrL0KxB:-Ny7gH-e36hk2xrL0KxC:bwhbn8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/builtins/pgvector_builtins.go#L41)):*
> ```Go
> 			},
> 			ReturnType: tree.FixedReturnType(types.Float),
> 			Fn: func(_ context.Context, evalCtx *eval.Context, args tree.Datums) (tree.Datum, error) {
> ```

These seem like good candidates to be defined with a SQL body using the corresponding operator. They're immutable, so would be inlined. What do you think?

Note: we'd also have to set `CalledOnNullInput: true` to freely allow inlining.
___
*[`pkg/sql/sem/cast/cast.go` line 254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7j9-i5w-Wx5Hmakhg:-Ny7j9-i5w-Wx5Hmakhh:b-3erzxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/cast/cast.go#L254)):*
> ```Go
> 	}
> 
> 	if srcFamily == types.ArrayFamily && tgtFamily == types.PGVectorFamily {
> ```

Should there also be an implicit cast from `vector` to `real[]` or `double[]`?

https://github.com/pgvector/pgvector/blob/258eaf58fdaff1843617ff59ea855e0768243fe9/sql/vector.sql#L157-L158
___
*[`pkg/sql/sem/eval/binary_op.go` line 1236 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7kWE23DJXYhdE_rEO:-Ny7kWE23DJXYhdE_rEP:b-7g0y0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/eval/binary_op.go#L1236)):*
> ```Go
> 	ctx context.Context, _ *tree.DistanceVectorOp, left, right tree.Datum,
> ) (tree.Datum, error) {
> 	v := tree.MustBeDPGVector(left)
> ```

I think this will result in an internal error if passed a NULL argument. Same for the builtin functions. We should also make sure to test that case.
___
*[`pkg/sql/sem/tree/datum.go` line 6673 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7oF6G2Vra7siKcOPQ:-Ny7oF6G2Vra7siKcOPR:b-fafv4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/tree/datum.go#L6673)):*
> ```Go
> 			width := int(typ.Width())
> 			if width > 0 && len(in.T) != width {
> 				return nil, pgerror.Newf(pgcode.StringDataLengthMismatch,
> ```

[nit] pg_vector uses `DataException`.
https://github.com/pgvector/pgvector/blob/258eaf58fdaff1843617ff59ea855e0768243fe9/src/vector.c#L77-L79
___
*[`pkg/sql/types/types.go` line 491 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7pydQ8hepoO2gYdqO:-Ny7pydQ8hepoO2gYdqP:b-rv2ryh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/types/types.go#L491)):*
> ```Go
> 
> 	// PGVector is the type representing a PGVector object.
> 	PGVector = &T{
> ```

Should we be adding a case to the `types.T.Equivalent()` method for `PGVector` since the widths of compatible vectors must be the same? (I think leaving it as-is could be justified as well)
___
*[`pkg/sql/types/types.go` line 1321 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7pRj4AYoZpJLiw3Bg:-Ny7pRj4AYoZpJLiw3Bh:b-zdrzth) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/types/types.go#L1321)):*
> ```Go
> //		COLLATEDSTRING: max # of characters
> //		BIT           : max # of bits
> //	  VECTOR        : # of dimensions
> ```

[nit] formatting
___
*[`pkg/util/vector/vector.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7saDmEw6w1A1eZaxm:-Ny7saDmEw6w1A1eZaxn:b3vas6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/util/vector/vector.go#L26)):*
> ```Go
> 
> // MaxDim is the maximum number of dimensions a vector can have.
> const MaxDim = 16000
> ```

Should we make this configurable? Even if only as a `crdb_internal` setting or env variable.
___
*[`pkg/util/vector/vector.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7xBLvAio162RD__qu:-Ny7xBLy8cuWWCw0xZ-3:byvl56m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/util/vector/vector.go#L55)):*
> ```Go
> 		}
> 
> 		val, err := strconv.ParseFloat(part, 32)
> ```

Should we also check for invalid floats? e.g. `NaN`, `+/-Inf`.
___
*[`pkg/util/vector/vector.go` line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7zHn3E_UPF-COd8Fc:-Ny7zHn3E_UPF-COd8Fd:bie3dea) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/util/vector/vector.go#L200)):*
> ```Go
> 		norm += float64(t[i]) * float64(t[i])
> 	}
> 	return math.Sqrt(norm)
> ```

We may want to validate the intermediate result here and for the l2 distance - we've run into flakes in aggregate functions that use `Sqrt` with floats when the true result is zero but precision errors result in a small negative value instead.

Also, are we missing range checks? This seems like it could overflow, as do some of the other functions.
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7z0BsCMANDQd03Npi:-Ny7z0Bt-8a9-tisnNui:b2mddqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L1)):*
> ```
> # LogicTest: !local-mixed-23.2
> ```

Can we do a test that attempts to pass a NULL value to some of the vector operators?
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 8 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7xjTK1ftNTR08jY5l:-Ny7xjTK1ftNTR08jY5m:b-6foe05) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L8)):*
> ```
> 5.196152422706632
> 
> statement error dimensions for type vector must be at least 1
> ```

[nit] We try to also validate the error code in logic tests with `pgcode <code>`
Ex: https://github.com/cockroachdb/cockroach/blob/e60b492dcbb2be398523cfddd9dae75e16f7cae1/pkg/sql/logictest/testdata/logic_test/udf#L15
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7yGrW7NnTj5W9P5gn:-Ny7yGrW7NnTj5W9P5go:b-einr3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L25)):*
> ```
> 
> query T rowsort
> SELECT * FROM v
> ```

It seems strange that we would allow this - aren't basically all vector operations impossible if the widths of the vectors are different?
___
*[`pkg/sql/rowenc/keyside/keyside_test.go` line 249 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny83J0ACzG4WW8vtZ5y:-Ny83J0ACzG4WW8vtZ5z:b-edrack) ([raw file](https://github.com/cockroachdb/cockroach/blob/76613277687c30c3345b9598c16ed6a350823988/pkg/sql/rowenc/keyside/keyside_test.go#L249)):*
> ```Go
> 		return hasKeyEncoding(typ.ArrayContents())
> 	}
> 	return !colinfo.MustBeValueEncoded(typ)
> ```

I think this change misses decimals, which also do not round trip in all cases.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
Thanks for the review! I'll just respond to a few of your astute questions for now before making modifications to the PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-:-Ny84aApCyqQnXZbofk_:b-hhjfz3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @nameisbhaskar, @nkodali, @rharding6373, @vidit-bhat, and @yuzefovich)

___
*[`pkg/sql/opt/ops/scalar.opt` line 538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7hBdH3chbzT3cCown:-Ny84j2X5BkmHuaT2BQu:b-npxtr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/opt/ops/scalar.opt#L538)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Any particular reason why we aren't also doing the l1 distance `<+>`? pg_vector seems to support it: https://github.com/pgvector/pgvector/blob/258eaf58fdaff1843617ff59ea855e0768243fe9/README.md?plain=1#L142-L147
</blockquote></details>

This was added in 0.7.0, and I started the patch when pgvector was still at 0.6.0. We could definitely add it, I was thinking of just adding it in a followon, but we could definitely do it now too.
___
*[`pkg/util/vector/vector.go` line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7zHn3E_UPF-COd8Fc:-Ny85VER8rhqtYXY-Dxi:bp4vbcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/util/vector/vector.go#L200)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We may want to validate the intermediate result here and for the l2 distance - we've run into flakes in aggregate functions that use `Sqrt` with floats when the true result is zero but precision errors result in a small negative value instead.

Also, are we missing range checks? This seems like it could overflow, as do some of the other functions.
</blockquote></details>

I tried to follow the pg_vector implementation, but if you think we need to add things like this I'm open to it. https://github.com/pgvector/pgvector/blob/master/src/vector.c
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7yGrW7NnTj5W9P5gn:-Ny84yTC9uPzAl0dHGTQ:b-tj0fy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L25)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It seems strange that we would allow this - aren't basically all vector operations impossible if the widths of the vectors are different?
</blockquote></details>

I actually just added this comparison today when responding to test failures, specifically because stats collection was panicking because it needs to run comparisons on the datums in a table column, and it's possible in pg_vector (for some reason) to have a vector column without a dimensions modifier. I checked the pg_vector source code and it also permits this comparison, so I figured I'd just throw it in here.

Thoughts?
___
*[`pkg/sql/rowenc/keyside/keyside_test.go` line 249 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny83J0ACzG4WW8vtZ5y:-Ny85Jjg4saBAZnbh2K1:bwxbmvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/76613277687c30c3345b9598c16ed6a350823988/pkg/sql/rowenc/keyside/keyside_test.go#L249)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think this change misses decimals, which also do not round trip in all cases.
</blockquote></details>

Good point, this wasn't failing any tests but I think you're right, will fix.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @nameisbhaskar, @nkodali, @rharding6373, @vidit-bhat, and @yuzefovich)

___
*[`pkg/sql/opt/ops/scalar.opt` line 538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7hBdH3chbzT3cCown:-Ny8FhaECi9F9x9iGI7c:b-pavbfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/opt/ops/scalar.opt#L538)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This was added in 0.7.0, and I started the patch when pgvector was still at 0.6.0. We could definitely add it, I was thinking of just adding it in a followon, but we could definitely do it now too.
</blockquote></details>

I think it'd be fine to leave it for a future PR, just curious.
___
*[`pkg/util/vector/vector.go` line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7zHn3E_UPF-COd8Fc:-Ny8Fqqw5rslMJ1GGtqg:b-9awdss) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/util/vector/vector.go#L200)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I tried to follow the pg_vector implementation, but if you think we need to add things like this I'm open to it. https://github.com/pgvector/pgvector/blob/master/src/vector.c
</blockquote></details>

I think it's ok to defer both requests (maybe indefinitely) until a test fails because of them, considering pg_vector does the exact same thing. But [nit] it might be nice to leave a TODO as a hint for the future.
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7yGrW7NnTj5W9P5gn:-Ny8KTRVDTV8NXMyMwcJ:babsm13) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L25)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I actually just added this comparison today when responding to test failures, specifically because stats collection was panicking because it needs to run comparisons on the datums in a table column, and it's possible in pg_vector (for some reason) to have a vector column without a dimensions modifier. I checked the pg_vector source code and it also permits this comparison, so I figured I'd just throw it in here.

Thoughts?
</blockquote></details>

Did you end up checking against a postgres instance with pg_vector installed? I think [this function](https://github.com/pgvector/pgvector/blob/258eaf58fdaff1843617ff59ea855e0768243fe9/src/vector.c#L366-L369) _might_ prevent table columns from being defined without an explicit dimension. At any rate, there aren't any pg_vector tests for this case, which makes it seem likely to be an oversight if they do allow it.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @nameisbhaskar, @nkodali, @rharding6373, @vidit-bhat, and @yuzefovich)

___
*[`pkg/sql/parser/sql.y` line 14277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7baQNA3H3KC7vw6Qf:-Ny8CKS2CGNNLWQ-pTFW:bm10ku5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/parser/sql.y#L14277)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can we add some parser tests?
</blockquote></details>

Yeah, I wasn't sure about needing to do this because all of this is implicitly tested by the logic tests and so on. The parser tests seem to be more used for new top level syntax, rather than new expression, type or operator syntax. Anyway, added.
___
*[`pkg/sql/scanner/scan.go` line 338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7e11DDpYZKMoTxH87:-Ny8DZyYAAvi_crzkME_:bj3uawm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/scanner/scan.go#L338)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Here and below, we're advancing the position of the scanner without setting the token ID. Should we backtrack if it isn't a vector operator?
</blockquote></details>

Yeah, good point. Fixed and added tests. The <- case threw me for a bit, but it scans because it could mean less than negative.
___
*[`pkg/sql/sem/builtins/pgvector_builtins.go` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7gH-e36hk2xrL0KxB:-Ny8KbZO8Shq0PtMFPbA:b8dtmo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/builtins/pgvector_builtins.go#L41)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

These seem like good candidates to be defined with a SQL body using the corresponding operator. They're immutable, so would be inlined. What do you think?

Note: we'd also have to set `CalledOnNullInput: true` to freely allow inlining.
</blockquote></details>

I don't really see the advantage beside code reuse, but if you feel strongly we can do it. I think a disadvantage is that the reader has to remember the weird names of the operators (`<=>` in this case) to validate that the code makes sense, whereas in the current implementation it's clear that cosine_distance calling CosDistance is the right thing.
___
*[`pkg/sql/sem/cast/cast.go` line 254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7j9-i5w-Wx5Hmakhg:-Ny8N35W7UqPytsXRYai:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/cast/cast.go#L254)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should there also be an implicit cast from `vector` to `real[]` or `double[]`?

https://github.com/pgvector/pgvector/blob/258eaf58fdaff1843617ff59ea855e0768243fe9/sql/vector.sql#L157-L158
</blockquote></details>

Done.
___
*[`pkg/sql/sem/eval/binary_op.go` line 1236 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7kWE23DJXYhdE_rEO:-Ny8NERX5BkxissdA9rd:b-cvwby2) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/eval/binary_op.go#L1236)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think this will result in an internal error if passed a NULL argument. Same for the builtin functions. We should also make sure to test that case.
</blockquote></details>

I think these should be fine because of logic that nullifies binary operators and builtins before they're called. I added tests, good point.
___
*[`pkg/sql/sem/tree/datum.go` line 6673 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7oF6G2Vra7siKcOPQ:-Ny8TUCh0vgP3g_6wOdT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/tree/datum.go#L6673)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] pg_vector uses `DataException`.
https://github.com/pgvector/pgvector/blob/258eaf58fdaff1843617ff59ea855e0768243fe9/src/vector.c#L77-L79
</blockquote></details>

Done.
___
*[`pkg/sql/types/types.go` line 491 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7pydQ8hepoO2gYdqO:-Ny8TmK896jUokaTbqJQ:b-in8n9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/types/types.go#L491)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we be adding a case to the `types.T.Equivalent()` method for `PGVector` since the widths of compatible vectors must be the same? (I think leaving it as-is could be justified as well)
</blockquote></details>

Hmm, but at least according to the pg_vector source code it is okay to compare vectors with different widths... 
___
*[`pkg/sql/types/types.go` line 1321 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7pRj4AYoZpJLiw3Bg:-Ny8Tepv8sjp4UbjpjGz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/types/types.go#L1321)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] formatting
</blockquote></details>

Done.
___
*[`pkg/util/vector/vector.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7saDmEw6w1A1eZaxm:-Ny8UerN3LHmuv7xyfJs:b-z9vtjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/util/vector/vector.go#L26)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we make this configurable? Even if only as a `crdb_internal` setting or env variable.
</blockquote></details>

Hmm... maybe? I don't see the use case. I think we should aim to stay completely compatible.
___
*[`pkg/util/vector/vector.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7xBLvAio162RD__qu:-Ny8Vj4Z6PC4c8qq3HC7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/util/vector/vector.go#L55)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we also check for invalid floats? e.g. `NaN`, `+/-Inf`.
</blockquote></details>

Done.
___
*[`pkg/util/vector/vector.go` line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7zHn3E_UPF-COd8Fc:-Ny8WN2V1UV75YLPg1GO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/util/vector/vector.go#L200)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think it's ok to defer both requests (maybe indefinitely) until a test fails because of them, considering pg_vector does the exact same thing. But [nit] it might be nice to leave a TODO as a hint for the future.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7z0BsCMANDQd03Npi:-Ny8WNZt1NhYnZ3tFnCV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L1)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can we do a test that attempts to pass a NULL value to some of the vector operators?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 8 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7xjTK1ftNTR08jY5l:-Ny8WOneCnWmVXse0ieT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L8)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] We try to also validate the error code in logic tests with `pgcode <code>`
Ex: https://github.com/cockroachdb/cockroach/blob/e60b492dcbb2be398523cfddd9dae75e16f7cae1/pkg/sql/logictest/testdata/logic_test/udf#L15
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7yGrW7NnTj5W9P5gn:-Ny8WRM-0W8YMBlyXmtz:bw2qszn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L25)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Did you end up checking against a postgres instance with pg_vector installed? I think [this function](https://github.com/pgvector/pgvector/blob/258eaf58fdaff1843617ff59ea855e0768243fe9/src/vector.c#L366-L369) _might_ prevent table columns from being defined without an explicit dimension. At any rate, there aren't any pg_vector tests for this case, which makes it seem likely to be an oversight if they do allow it.
</blockquote></details>

I did, it is permitted :/

```
jordan=# create table t(v vector);
CREATE TABLE
jordan=# insert into t values('[1]'),('[1,2]');
INSERT 0 2
jordan=# select * from t;
   v
-------
 [1]
 [1,2]
(2 rows)
```


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
This is great work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once you add a case to `unsupported_types.go`. I think Yahor will be taking a look as well.

> Reviewed 14 of 14 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-:-NyFAxhA7K5WmN5AN0qh:b-o28hv5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @nameisbhaskar, @nkodali, @rharding6373, @vidit-bhat, and @yuzefovich)

___
*[`pkg/sql/sem/builtins/pgvector_builtins.go` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7gH-e36hk2xrL0KxB:-NyF2y1a3997Gz5lfD5h:br7ww9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/builtins/pgvector_builtins.go#L41)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't really see the advantage beside code reuse, but if you feel strongly we can do it. I think a disadvantage is that the reader has to remember the weird names of the operators (`<=>` in this case) to validate that the code makes sense, whereas in the current implementation it's clear that cosine_distance calling CosDistance is the right thing.
</blockquote></details>

My thinking was that it'd be nicer for the optimizer and vectorized engine (when we add a vectorized implementation) to only have to consider the operators. This is something we could change later, though, so I'd be alright with leaving it this way for now if you prefer. I guess it would also be pretty simple to just add a couple norm rules that convert the builtin function calls to the corresponding operators.
___
*[`pkg/sql/sem/eval/binary_op.go` line 1236 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7kWE23DJXYhdE_rEO:-NyF6ZzC0yq_56Va6F3m:b-u319d) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/sem/eval/binary_op.go#L1236)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think these should be fine because of logic that nullifies binary operators and builtins before they're called. I added tests, good point.
</blockquote></details>

TIL`BinOp` also has a `CalledOnNullInput` field.
___
*[`pkg/sql/types/types.go` line 491 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7pydQ8hepoO2gYdqO:-NyF7xTSAIoGvI3wd7_f:b-ntibar) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/types/types.go#L491)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm, but at least according to the pg_vector source code it is okay to compare vectors with different widths...
</blockquote></details>

Right, no need for a change then.
___
*[`pkg/sql/logictest/testdata/logic_test/mixed_version_pgvector` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NyF0vSVEX8Nd4JYerO0:-NyF0vSW69FpnLWUZsz1:b-5rz482) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/logictest/testdata/logic_test/mixed_version_pgvector#L1)):*
> ```
> # LogicTest: cockroach-go-testserver-23.2
> ```

Thanks for adding this. It reminds me - we also need a case for `VECTOR` in `unsupported_types.go`.
___
*[`pkg/sql/logictest/testdata/logic_test/vector` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-Ny7yGrW7NnTj5W9P5gn:-NyEz4OA8o13hTo5Ndv9:bhrf8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa290d760a0bd1316e1aae4c0d6e1099fff0e5f9/pkg/sql/logictest/testdata/logic_test/vector#L25)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I did, it is permitted :/

```
jordan=# create table t(v vector);
CREATE TABLE
jordan=# insert into t values('[1]'),('[1,2]');
INSERT 0 2
jordan=# select * from t;
   v
-------
 [1]
 [1,2]
(2 rows)
```
</blockquote></details>

Ah well, that's a little sad. Thanks for checking.
___
*[`pkg/sql/parser/testdata/select_exprs` line 2068 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NyF9XVK-Z1-JYEayh4L:-NyF9XVK-Z1-JYEayh4M:b-o8g4pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/parser/testdata/select_exprs#L2068)):*
> ```
> SELECT (("[1,2]") <-> ("[3,4]")) -- fully parenthesized
> SELECT "[1,2]" <-> "[3,4]" -- literals removed
> SELECT _ <-> _ -- identifiers removed
> ```

Do you know why the vectors are considered identifiers instead of literals? Just because of the double quotes?


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
pgvector 0.7 is brand new, but I still think it's worth getting some of the new functionality in for this initial version. In particular the new `halfvec` type is significant (it's necessary to support openai's larger embedding model without truncation) and we should make sure we have the right structure for multiple vector types. 

One thing that pgvector doesn't do AFAICT but seems useful is if each vector had a bit to indicate whether it is normalized, so that cosine_distance could automatically turn into the faster inner_product when we don't need to normalize again. Would that be too much of a deviation from pgvector?

I was thinking it would be best to treat vectors as binary blobs as much as possible, and only view them as arrays when needed. That would minimize decoding overhead on the assumption that eventually we're not computing inner products with a Go for loop but passing the blob off to some simd-optimized implementation. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-:-NyVnrjh4PaYnfFBNhR8:b-5karjb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @jordanlewis, @nameisbhaskar, @nkodali, @rharding6373, @vidit-bhat, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Great stuff! I only have some minor nits (and I'll keep on learning about the pgvector implementation).

> Reviewed 51 of 68 files at r1, 12 of 21 files at r2, 2 of 4 files at r3, 14 of 14 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-:-Nyc_onM3-G3JL45cZlO:b-476umj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @jordanlewis, @nameisbhaskar, @nkodali, @rharding6373, and @vidit-bhat)

___
*[`pkg/sql/parser/sql.y` line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NyR3ZixD4rafjHNXP2N:-NyR3ZixD4rafjHNXP2O:b-x877hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/parser/sql.y#L34)):*
> ```Yacc
> 
>     "github.com/cockroachdb/cockroach/pkg/geo/geopb"
>     "github.com/cockroachdb/cockroach/pkg/geo/geopb"
> ```

nit: this import is duplicated with the line above.
___
*[`pkg/sql/randgen/datum.go` line 297 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NyR4XMZ7gc-1J9zEaMe:-NyR4XMZ7gc-1J9zEaMf:b58wkqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/randgen/datum.go#L297)):*
> ```Go
> 	case types.TSQueryFamily:
> 		return tree.NewDTSQuery(tsearch.RandomTSQuery(rng))
> 	case types.PGVectorFamily:
> ```

nit: are there some interesting vectors we could add to `randInterestingDatums`?

Relatedly, is PGVector a scalar type? In other words, should it be included into `types.ScalarTypes`?
___
*[`pkg/sql/rowenc/encoded_datum.go` line 333 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NyR51xs9X1AOxYuVkrF:-NyR51xs9X1AOxYuVkrG:bilougi) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/rowenc/encoded_datum.go#L333)):*
> ```Go
> 		// Note that at time of this writing we don't support arrays of JSON
> 		// (tracked via #23468) nor of TSQuery / TSVector / PGVector types (tracked by
> 		// #90886), so technically we don't need to do a recursive call here,
> ```

nit: arrays of PGVector type are tracked by #121432.
___
*[`pkg/sql/sem/cast/cast_map.go` line 75 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycRV2V3VAiY_FNCigm:-NycRV2V3VAiY_FNCign:b-j3xbk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/sem/cast/cast_map.go#L75)):*
> ```Go
> 		oid.T_text:    {MaxContext: ContextAssignment, origin: ContextOriginAutomaticIOConversion, Volatility: volatility.Immutable},
> 	},
> 	oidext.T_pgvector: {
> ```

Just checking: updates to `castMap` were generated via `cast_map_gen.sh`?
___
*[`pkg/util/vector/vector.go` line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycWqZG9TqCfeKSQv56:-NycWqZG9TqCfeKSQv57:b-i8y16i) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/util/vector/vector.go#L74)):*
> ```Go
> // String implements the fmt.Stringer interface.
> func (v T) String() string {
> 	strs := make([]string, len(v))
> ```

nit: should we use `strings.Builder`?
___
*[`pkg/util/vector/vector.go` line 83 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycWz0rDbOuCTqnUZ7W:-NycWz0sAHE9X8LRIbDh:bkgfvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/util/vector/vector.go#L83)):*
> ```Go
> // Size returns the size of the vector in bytes.
> func (v T) Size() uintptr {
> 	return uintptr(len(v)) * 4
> ```

nit: should we do `s/len/cap/` to be more precise about memory usage? Also perhaps include 24 bytes for the slice overhead?
___
*[`pkg/util/vector/vector.go` line 175 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycYDUv1lKj7GdGILoW:-NycYDUwDVfpdoNwe5Nq:b-q3vvuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/util/vector/vector.go#L175)):*
> ```Go
> 		normB += t2[i] * t2[i]
> 	}
> 	// Use sqrt(a * b) over sqrt(a) * sqrt(b)
> ```

nit: I see that this comment comes from the pgvector source code, but I find it confusing since it doesn't actually match the formula.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @nameisbhaskar, @nkodali, @rharding6373, @vidit-bhat, and @yuzefovich)

___
*[`pkg/sql/randgen/datum.go` line 297 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NyR4XMZ7gc-1J9zEaMe:-O0IynvW7dYTKzkFNdsK:bfv3dv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/randgen/datum.go#L297)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: are there some interesting vectors we could add to `randInterestingDatums`?

Relatedly, is PGVector a scalar type? In other words, should it be included into `types.ScalarTypes`?
</blockquote></details>

I don't think so because the length information is important but the interesting datums code doesn't have a hook for that. I guess we could add it, but I'm not sure how important this is.

Added to scalar types.
___
*[`pkg/sql/sem/cast/cast_map.go` line 75 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycRV2V3VAiY_FNCigm:-O0Iw2nbCgZnxzEOQeny:b-x6lp3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/sem/cast/cast_map.go#L75)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Just checking: updates to `castMap` were generated via `cast_map_gen.sh`?
</blockquote></details>

No, this script doesn't seem to create pgvector outputs.
___
*[`pkg/util/vector/vector.go` line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycWqZG9TqCfeKSQv56:-O0IxKoM1uijwG4SHL9c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/util/vector/vector.go#L74)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we use `strings.Builder`?
</blockquote></details>

Done.
___
*[`pkg/util/vector/vector.go` line 83 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycWz0rDbOuCTqnUZ7W:-O0IxJm64jtkAT61n-Lv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/util/vector/vector.go#L83)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we do `s/len/cap/` to be more precise about memory usage? Also perhaps include 24 bytes for the slice overhead?
</blockquote></details>

Done.
___
*[`pkg/util/vector/vector.go` line 175 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycYDUv1lKj7GdGILoW:-O0IxL_f28bUybgNUYc2:b-hxmzhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/util/vector/vector.go#L175)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I see that this comment comes from the pgvector source code, but I find it confusing since it doesn't actually match the formula.
</blockquote></details>

I think the comment says "over" but it really means "instead of"?
___
*[`pkg/sql/logictest/testdata/logic_test/mixed_version_pgvector` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NyF0vSVEX8Nd4JYerO0:-O0IxReN3OBuK-OfynlV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/sql/logictest/testdata/logic_test/mixed_version_pgvector#L1)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Thanks for adding this. It reminds me - we also need a case for `VECTOR` in `unsupported_types.go`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
> Added to scalar types.

I removed it from scalar types again, since that causes `max` and `min` to be generated which aren't actually supposed to be generated. I don't think it's actually a scalar type - even though it's a primitive it's essentially just an array.

jordanlewis commented:
I'd like to get this merged, at the very least to stop rebasing the `fixed_oids` file :) Any objections? The current status is that I've responded to all comments, but haven't added any of Ben's suggestions.

yuzefovich reviewed (approved):
No objections from me, ship it! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 21 files at r2, 29 of 38 files at r5, 2 of 3 files at r6, 7 of 7 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-:-O0o6fnGDCdG76YEtMzV:bjgonn8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @jordanlewis, @nameisbhaskar, @nkodali, @rharding6373, and @vidit-bhat)

___
*[`pkg/util/vector/vector.go` line 175 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-NycYDUv1lKj7GdGILoW:-O0o2wtg1QnXrNxxd8LT:b-qfaanf) ([raw file](https://github.com/cockroachdb/cockroach/blob/76ad36ddc14d993bd9738e2b2fd8fbe7a2e1b14c/pkg/util/vector/vector.go#L175)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think the comment says "over" but it really means "instead of"?
</blockquote></details>

Ah, I see, thanks.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
Looks like there was a test failure:
```
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17858_17858&logFilter=debug&logView=flowAware)      --- FAIL: BenchmarkParquetWriter/vector
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17859_17859&logFilter=debug&logView=flowAware)          writer_bench_test.go:44:
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17860_17860&logFilter=debug&logView=flowAware)                Error Trace:  github.com/cockroachdb/cockroach/pkg/util/parquet/writer_bench_test.go:44
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17861_17861&logFilter=debug&logView=flowAware)                                    GOROOT/src/testing/benchmark.go:193
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17862_17862&logFilter=debug&logView=flowAware)                                    GOROOT/src/testing/benchmark.go:215
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17863_17863&logFilter=debug&logView=flowAware)                                    src/runtime/asm_amd64.s:1695
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17864_17864&logFilter=debug&logView=flowAware)                Error:        Received unexpected error:
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17865_17865&logFilter=debug&logView=flowAware)                              parquet writer does not support the type family PGVectorFamily
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17866_17866&logFilter=debug&logView=flowAware)                              (1) candidate pg code: 0A000
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17867_17867&logFilter=debug&logView=flowAware)                              Wraps: (2) attached stack trace
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17868_17868&logFilter=debug&logView=flowAware)                                -- stack trace:
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17869_17869&logFilter=debug&logView=flowAware)                                | github.com/cockroachdb/cockroach/pkg/util/parquet.makeColumn
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17870_17870&logFilter=debug&logView=flowAware)                                |   github.com/cockroachdb/cockroach/pkg/util/parquet/schema.go:483
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17871_17871&logFilter=debug&logView=flowAware)                                | github.com/cockroachdb/cockroach/pkg/util/parquet.NewSchema
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17872_17872&logFilter=debug&logView=flowAware)                                |   github.com/cockroachdb/cockroach/pkg/util/parquet/schema.go:96
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17873_17873&logFilter=debug&logView=flowAware)                                | github.com/cockroachdb/cockroach/pkg/util/parquet.BenchmarkParquetWriter.func1
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17874_17874&logFilter=debug&logView=flowAware)                                |   github.com/cockroachdb/cockroach/pkg/util/parquet/writer_bench_test.go:43
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17875_17875&logFilter=debug&logView=flowAware)                                | testing.(*B).runN
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17876_17876&logFilter=debug&logView=flowAware)                                |   GOROOT/src/testing/benchmark.go:193
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17877_17877&logFilter=debug&logView=flowAware)                                | testing.(*B).run1.func1
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17878_17878&logFilter=debug&logView=flowAware)                                |   GOROOT/src/testing/benchmark.go:215
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17879_17879&logFilter=debug&logView=flowAware)                                | runtime.goexit
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17880_17880&logFilter=debug&logView=flowAware)                                |   src/runtime/asm_amd64.s:1695
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17881_17881&logFilter=debug&logView=flowAware)                              Wraps: (3) parquet writer does not support the type family PGVectorFamily
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17882_17882&logFilter=debug&logView=flowAware)                              Error types: (1) *pgerror.withCandidateCode (2) *withstack.withStack (3) *errutil.leafError
[04:49:14 ](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/15897801?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=15897799_17883_17883&logFilter=debug&logView=flowAware)                Test:         BenchmarkParquetWriter/vector
```

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 29 of 38 files at r5, 2 of 3 files at r6, 7 of 7 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124292#-:-O0pGvPv0XKVq3BArKH2:bcmw7ss)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @jordanlewis, @nameisbhaskar, @nkodali, @rharding6373, and @vidit-bhat)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
Latest round:

- Removed Avro support based on @asg0451's suggestion. 
- Fixed Parquet benchmark failure (Parquet also not supported).
- Moved the vector datatype into enterprise based on @dikshant's comment.

No substantive changes were needed to accomplish this so I'm going to merge it.

bors r+
